### ISSUE TITLE
Implement CoSENT loss

### ISSUE THREAD

**Issue by tomaarsen** [UKPLab/sentence-transformers#2438]
Hello!

## Context
This issue describes a feature that I am planning to be included in a release before v3, or alternatively, in v3 of Sentence Transformers.

## Details
Some [recent work by Li & Li, 2023](https://arxiv.org/abs/2309.12871) brought my attention to CoSENT ([blogpost](https://kexue.fm/archives/8847), [GitHub repo](https://github.com/bojone/CoSENT)). It proposes a new loss function with the purpose of more effectively and directly causing the cosine similarity of a positive pair to be larger than that of a negative pair.

Implementation snippets: 
* in Keras: https://github.com/bojone/CoSENT/blob/124c368efc8a4b179469be99cb6e62e1f2949d39/cosent.py#L79-L90
* in Torch: https://github.com/SeanLee97/AnglE/blob/ce6d402c794fe3ddd1c1c08a756124cd04f6bb65/angle_emb/angle.py#L67-L77

@johneckberg has expressed an interest in contributing this loss.

cc @bwanglzu @ir2718 @johneckberg @aamir-s18 as I know you're interested in my TODO list. John will lead this development, and I will review his PR.

- Tom Aarsen
